### ISSUE TITLE
fortio mtls setup question

### ISSUE THREAD

**Issue by eoregua** [fortio/fortio#530]
Hi,

Is it supported mutual TLS from fortio client?

In my test setup environment when the server (envoy) is configured to require the client certificate (mutual TLS) the TLS handshake is failing:

e.g:

```fortio load -a -grpc -ping -cacert cacert.pem -qps 1000 -c 2 -s 4 tls-p_local.it:31136
Fortio 1.21.2 running at 1000 queries per second, 4->4 procs, for 5s: tls-p_local.it:31136
08:58:06 I grpcrunner.go:153> Starting GRPC Ping test for tls-p_local.it:31136 with 4*2 threads at 1000.0 qps
08:58:06 I grpcrunner.go:49> Using CA certificate cacert.pem to construct TLS credentials
08:58:06 E grpcrunner.go:202> Error in first grpc call (ping = true) for tls-p_local.it:31136: rpc error: code = Unavailable desc = connection closed before server preface received
Aborting because of rpc error: code = Unavailable desc = connection closed before server preface received
```

If I change the configuration in envoy (require_client_certificate: false -> simple TLS) the TLS handshake and the load test using grpc ping is correctly done.

e.g:

```
fortio load -a -grpc -ping -cacert cacert.pem -qps 1000 -c 2 -s 4 tls-p_local.it:31136
Fortio 1.21.2 running at 1000 queries per second, 4->4 procs, for 5s: tls-p_local.it:31136
08:39:03 I grpcrunner.go:153> Starting GRPC Ping test for tls-p_local.it:31136 with 4*2 threads at 1000.0 qps
08:39:03 I grpcrunner.go:49> Using CA certificate cacert.pem to construct TLS credentials
08:39:03 I grpcrunner.go:49> Using CA certificate cacert.pem to construct TLS credentials
Starting at 1000 qps with 8 thread(s) [gomax 4] for 5s : 625 calls each (total 5000)
08:39:08 I periodic.go:693> T001 ended after 5.002394511s : 625 calls. qps=124.94016587969185
08:39:08 I periodic.go:693> T002 ended after 5.002416367s : 625 calls. qps=124.93962000504544
08:39:08 I periodic.go:693> T005 ended after 5.002443451s : 625 calls. qps=124.93894356268255
08:39:08 I periodic.go:693> T003 ended after 5.002454861s : 625 calls. qps=124.93865859192607
08:39:08 I periodic.go:693> T007 ended after 5.002465836s : 625 calls. qps=124.93838448675015
08:39:08 I periodic.go:693> T000 ended after 5.002463639s : 625 calls. qps=124.93843935763988
08:39:08 I periodic.go:693> T004 ended after 5.002461657s : 625 calls. qps=124.93848885886624
08:39:08 I periodic.go:693> T006 ended after 5.002480705s : 625 calls. qps=124.93801312922805
Ended after 5.002634956s : 5000 calls. qps=999.47
Sleep times : count 4992 avg 0.0062652854 +/- 0.0006128 min -0.002275793 max 0.007185069 sum 31.2763049
Aggregated Function Time : count 5000 avg 0.0011247684 +/- 0.0003316 min 0.000567061 max 0.010020304 sum 5.62384191
# range, mid point, percentile, count
>= 0.000567061 <= 0.001 , 0.000783531 , 22.46, 1123
> 0.001 <= 0.002 , 0.0015 , 98.68, 3811
> 0.002 <= 0.003 , 0.0025 , 99.68, 50
> 0.003 <= 0.004 , 0.0035 , 99.92, 12
> 0.009 <= 0.01 , 0.0095 , 99.96, 2
> 0.01 <= 0.0100203 , 0.0100102 , 100.00, 2
# target 50% 0.00136132
# target 75% 0.00168932
# target 90% 0.00188612
# target 99% 0.00232
# target 99.9% 0.00391667
Jitter: false
Ping SERVING : 5000
All done 5000 calls (plus 2 warmup) 1.125 ms avg, 999.5 qps
```
Can you please clarify if it's an issue/limitation or fortio load should be run with differents options to run over mutual TLS?

Thanks!

edited to use ```

**Comment by ldemailly** [comment]
For mTLS ie mutual TLS the client need to send its certificate, so you need to use `-cert` and `-key` not just `-cacert`

**Comment by ldemailly** [comment]
Will close assuming I answered your question, do reopen if that doesn't work/you have further questions

**Comment by eoregua** [comment]
I tried again but always the same error:

```
fortio load -a -grpc -ping -cert client.pem -key clientkey.pem -cacert cacert.pem  -qps 100 -c 2 -t 64s tls-p_local.it:31136
Fortio 1.21.2 running at 100 queries per second, 4->4 procs, for 1m4s: tls-p_local.it:31136
08:47:48 I grpcrunner.go:153> Starting GRPC Ping test for tls-p_local.it:31136 with 1*2 threads at 100.0 qps
08:47:48 I grpcrunner.go:49> Using CA certificate cacert.pem to construct TLS credentials
08:47:48 E grpcrunner.go:202> Error in first grpc call (ping = true) for tls-p_local.it:31136: rpc error: code = Unavailable desc = connection closed before server preface received
Aborting because of rpc error: code = Unavailable desc = connection closed before server preface received
```

**Comment by ldemailly** [comment]
I realized #141 is actually not done done for grpc

just to confirm if you remove `-grpc` and put `https://tls-p_local.it:31136/` as your url does it work?

**Comment by ldemailly** [comment]
should be fixed in `tls_options_sharing` branch can you confirm? tagged as `v1.24.0-pre1`

**Comment by eoregua** [comment]
Yes, from a quick check seems that with the new build mTLS is working fine now with grpc:
```
fortio load -a -grpc -ping -cert server.pem -key serverkey.pem -cacert cacert.pem  -qps 100 -c 2 -t 64s tls-p_local.it:31136
Fortio 1.24.0-pre1 running at 100 queries per second, 4->4 procs, for 1m4s: tls-p_local.it:31136
19:56:17 I grpcrunner.go:145> Starting GRPC Ping test for tls-p_local.it:31136 with 1*2 threads at 100.0 qps
Starting at 100 qps with 2 thread(s) [gomax 4] for 1m4s : 3200 calls each (total 6400)
19:57:21 I periodic.go:693> T001 ended after 1m4.002362068s : 3200 calls. qps=49.99815470247997
19:57:21 I periodic.go:693> T000 ended after 1m4.002371533s : 3200 calls. qps=49.99814730849561
Ended after 1m4.002418206s : 6400 calls. qps=99.996
Sleep times : count 6398 avg 0.018049786 +/- 0.0005477 min 0.001223931 max 0.019301548 sum 115.482532
Aggregated Function Time : count 6400 avg 0.0013553249 +/- 0.0002645 min 0.000667089 max 0.007775556 sum 8.67407937
# range, mid point, percentile, count
>= 0.000667089 <= 0.001 , 0.000833545 , 0.72, 46
> 0.001 <= 0.002 , 0.0015 , 99.09, 6296
> 0.002 <= 0.003 , 0.0025 , 99.66, 36
> 0.003 <= 0.004 , 0.0035 , 99.88, 14
> 0.004 <= 0.005 , 0.0045 , 99.89, 1
> 0.005 <= 0.006 , 0.0055 , 99.92, 2
> 0.006 <= 0.007 , 0.0065 , 99.95, 2
> 0.007 <= 0.00777556 , 0.00738778 , 100.00, 3
# target 50% 0.00150095
# target 75% 0.00175508
# target 90% 0.00190756
# target 99% 0.00199905
# target 99.9% 0.0053
Jitter: false
Ping SERVING : 6400
All done 6400 calls (plus 2 warmup) 1.355 ms avg, 100.0 qps
```
Thanks!

**Comment by ldemailly** [comment]
cheers! thanks!

**Comment by ldemailly** [comment]
fixed in (upcoming soon) 1.24.0
